### PR TITLE
BitmapData: implement applyFilter(), fixes #1068, full credit for solution goes to @PeyTy

### LINKED ISSUE #1068: Filters work, but not with BitmapData.draw() ($50 bounty)
``` haxe
package;


import flash.filters.GlowFilter;
import openfl.display.Bitmap;
import openfl.display.BitmapData;
import openfl.display.Sprite;
import openfl.Assets;


class Main extends Sprite {


    public function new () {

        super ();

        var bitmap = new Bitmap (Assets.getBitmapData ("assets/openfl.png"));
        addChild (bitmap);

        bitmap.x = (stage.stageWidth - bitmap.width) / 2;
        bitmap.y = (stage.stageHeight - bitmap.height) / 2;

        var filter = new GlowFilter();
        bitmap.filters = [filter];

        var bd2:BitmapData = new BitmapData(Std.int(bitmap.width), Std.int(bitmap.height));
        bd2.draw(bitmap);

        var bitmap2 = new Bitmap(bd2);
        addChild(bitmap2);
    }


}

```

Produces these results in neko:
2016-03-22 14_14_18-displaying a bitmap

Both images should have the filter applied to them, but only the first one (the one on the display list) shows it.

Also, this can't be compiled with flash? I keep getting the error "Recursive typedef is not allowed" for GlowFilter.hx, not sure what that's about.

Using latest development version of Lime & OpenFL.

### PR DESCRIPTION
**Pull request by larsiusprime**
also makes BitmapData.draw() filter-aware!

### CODE REVIEW

File: openfl/display/BitmapData.hx [hunk 1]
```diff
@@ -514,6 +531,16 @@ class BitmapData implements IBitmapDrawable {
 		
 		surface.flush ();
 		
+		var renderer = @:privateAccess Lib.current.stage.__renderer;
+		var renderSession = @:privateAccess renderer.renderSession;
+		__drawGL (renderSession, this);
+		
+		if (Std.is(source, DisplayObject) && cast(source, DisplayObject).filters != null) {
+			
+			@:privateAccess BitmapFilter.__applyFilters (cast(source, DisplayObject).filters, renderSession, this, this, null, null);
```
Comment (larsiusprime) [inline openfl/display/BitmapData.hx]: @PeyTy: was this line what you were referring to?
> P.S.  Forgot to add sourceRect, destPoint into BitmapFilter.__applyFilters in BD.draw call!

Comment (PeyTy) [inline openfl/display/BitmapData.hx]: Yes, but this makes no sense, sorry. Just got to home, was quick in decisions :)

**Comment by ibilon** [comment]
Is this ready? Couldn't seem to see the change when testing with your code from #1068.

**Comment by PeyTy** [comment]
Hello, Valentin! Should work most of the time. Did you checked my last
fixes from issue?
May you provide code sample?

**Comment by ibilon** [comment]
I only checked out this pr, and tested with https://github.com/openfl/openfl/issues/1068#issue-142752261 and https://github.com/openfl/openfl/issues/1068#issuecomment-212635044.

**Comment by larsiusprime** [comment]
@ibilon there's an edge case w/ flixel -- you can see it in the flxshapes demo, where filtered lightning turns out squished for some reason w/ this change. simple openfl demos seem fine.

**Comment by PeyTy** [comment]
I am not sure if Lars updated pull request. His samples should work
100%  (with latest workarounds applied), and 99% in other cases (some
uncommon behavior still buggy). Ask him to update PR.

суббота, 23 апреля 2016 г. пользователь Valentin Lemière написал:

**Comment by larsiusprime** [comment]
@PeyTy message received, I'll update this on Monday. Don't merge just yet.

**Pull request closed without merging**